File: src/y03.js
console.log ((function () {

  // Higher order functions
  // Functional Refatorings
  // * Tennent Correspondence Principle
  // * Inline function def
  // * Rescope Variable

  // GOAL: fact = FIND_FIXPOINT (fimp)

  error = function (n) { throw Error ("SHOULD NOT BE CALLED!") }

  fimp = function (fp) {
    return (function (n) {
      return n==0 ? 1 : n * fp (n-1)
    })
  }

  Y = function (f) {
    return (function (g) { return function (n) { return f (g (g)) (n) } } ) (
            function (g) { return function (n) { return f (g (g)) (n) } } ) }

  fact = Y (fimp)

  result = fact (5);

  return result;

}) ());

File: src/y03.clj
(println
 (do

   (def error (fn [n] (throw (Throwable. "SHOULD NEVER BE CALLED"))))

   (def fact-partial
        (fn [recursive]
          (fn [n] (if (zero? n) 1 (* n (recursive (- n 1)))))))

   (def y
        (fn [prf]
          ((fn [g] (fn [n] ((prf (g g)) n)))
           (fn [g] (fn [n] ((prf (g g)) n))))))

   (def fact
        (y fact-partial))

   (fact  5)

   ))
